### PR TITLE
Addition method testing with JUnit

### PR DESCRIPTION
**Pull request by julianmateu**
Is it OK to use assertTrue?, if not:
-Should I override the Equals method or define a new assertEquals?
-How do I override the Equals(Object object) method to use assertEquals in JUnit?
     http://stackoverflow.com/questions/6060848/junit-assertequals-fails-for-two-objects
     http://stackoverflow.com/questions/2316220/java-override-object-equals-method

### CODE REVIEW

File: src/Quaternion.java [hunk 1]
```diff
@@ -31,6 +68,19 @@ public Quaternion add(Quaternion addend) {
             this.k + addend.k);
     }
 
+    /** Method to assert if to Quaternions are equal.
+     * @return Returns True if they are equal, and False otherwise.
+     */
+    //@Override
+    public boolean equals(Quaternion compared) {
+
```
Comment (basicNew) [inline src/Quaternion.java:75]: should use getters and setters now that you have them :)

File: src/Quaternion.java [hunk 2]
```diff
@@ -31,6 +68,19 @@ public Quaternion add(Quaternion addend) {
...
+     * @return Returns True if they are equal, and False otherwise.
+     */
+    //@Override
+    public boolean equals(Quaternion compared) {
+        
+        if (this.r != compared.r) return false;
+        if (this.i != compared.i) return false;
+        if (this.j != compared.j) return false;
+        if (this.k != compared.k) return false;
+        else return true;
```
Comment (basicNew) [inline src/Quaternion.java]: Mostly a matter of style, but why not a single return using `AND`s?

File: src/TestQuaternion.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+public class TestQuaternion extends TestCase {
+
+    public TestQuaternion(String name) {
+        super(name);
+    }
+
+    public void testEquals() {
+        Quaternion a = new Quaternion(1,0,-1,5);
+
+        assertTrue(a.equals(new Quaternion(1,0,-1,5)));
```
Comment (basicNew) [inline src/TestQuaternion.java]: Maybe `assertEquals(...,...)` is better here to convey the idea

File: src/TestQuaternion.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    public void testAddition() {
+        Quaternion a = new Quaternion(1,1,1,-1);
+        Quaternion b = new Quaternion(1,0,-1,-1);
+        Quaternion result = new Quaternion(2,1,0,-2);
+
+        assertTrue(result.equals(a.add(b)));
```
Comment (basicNew) [inline src/TestQuaternion.java]: Again, something along the lines of `assertEquals(a.add(b), result)` is more readable and less code :)

**Comment by basicNew** [comment]
Looks good for a starter. The question now is if you can come up with any edge cases to add to your tests :)

File: src/TestQuaternion.java (hunk 3 above)
Comment (julianmateu) [inline src/TestQuaternion.java]: That was one of the questions I had, because I tried with `assertEquals(a,new Quaternion(1,0,-1,5));` and got a message saying:

```
There was 1 failure:
1) testEquals(TestQuaternion)junit.framework.AssertionFailedError: expected:<( 1.0 ; 0.0 ; -1.0 ; 5.0 )> but was:<( 1.0 ; 0.0 ; -1.0 ; 5.0 )>
```

I did some research (http://stackoverflow.com/questions/6060848/junit-assertequals-fails-for-two-objects), and found that when you define a method called `equals()` for a class you create, you should write one with the following header, because this is the one JUnit uses in `assertEquals()`:

```
@Override
public boolean equals(Object other)
```

However, when I tried this I came across the problem that a generic Object doesn't have the fields (r,i,j,k) nor the setters and getters.

I searched how to define an `equals(Object)` method that overrides the default one and found (http://stackoverflow.com/questions/2316220/java-override-object-equals-method) that you can do something like check that the object is the type you want and then cast it, but I got a little confused by the comments, and since I was certain that `assertTrue()` worked, I used that instead of `assertEquals()`.

Comment (basicNew) [inline src/TestQuaternion.java]: That's great, no problem! Let's discuss this next time we meet.
